epoll/src/main/scala/fs2/io/epoll/unsafe/EpollSystem.scala

```diff
@@ -0,0 +1,255 @@
+/*
+ * Copyright 2023 Arman Bilge
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package fs2.io.epoll
+package unsafe
+
+import cats.effect.{IO, Resource, FileDescriptorPoller, FileDescriptorPollHandle}
+import cats.effect.std.Mutex
+import cats.syntax.all._
+import cats.effect.unsafe.PollingSystem
+
+import java.io.IOException
+
+import jnr.ffi.Memory;
+import jnr.constants.platform.Errno;
+
+import scala.collection.mutable.HashMap
+import scala.annotation.tailrec
+
+object EpollSystem extends PollingSystem {
+
+  import epoll._
+  import libc.jnr._
+
+  private[this] final val MaxEvents = 64
+
+  type Api = FileDescriptorPoller
+
+  def close(): Unit = ()
+
+  def makeApi(register: (Poller => Unit) => Unit): Api =
+    new FileDescriptorPollerImpl(register)
+
+  def makePoller(): Poller = {
+    val fd = epoll_create1(0)
+    if (fd == -1)
+      throw new IOException(strerror(errno()))
+    new Poller(fd)
+  }
+
+  def closePoller(poller: Poller): Unit = poller.close()
+
+  def poll(poller: Poller, nanos: Long, reportFailure: Throwable => Unit): Boolean =
+    poller.poll(nanos)
+
+  def needsPoll(poller: Poller): Boolean = poller.needsPoll()
+
+  def interrupt(targetThread: Thread, targetPoller: Poller): Unit = ()
```

Comment:
Antonio is right, we definitely need interruption in multi-threading.

You can check how Antonio did it with a pipe, but I think in this case we should prefer to use an `eventfd`. An `eventfd` is kind of similar to FS2 `SignallingRef`: it is an atomic counter that you can wait for events on.

https://man7.org/linux/man-pages/man2/eventfd.2.html

The basic idea is:
1. We create an `eventfd` for each poller.
2. We should setup a read-ready event on the epoll for that `eventfd`. It will be read-ready whenever the value of the eventfd becomes greater than 0.
3. We can implement interruption for the poller by incrementing the value of the `eventfd` by writing to it. This will cause it to be greater than 0, and thus wake it up if it is polling.